File: src/services/liveNetworkCapabilities.ts
export const LIVE_NETWORK_PROTOCOL_VERSION = "hook.live.network.v1" as const;

export type LiveNetworkScope =
    | "loopback_http"
    | "loopback_https"
    | "private_https"
    | "unavailable";

export interface LiveNetworkTransportCapability {
    id: "websocket_binary" | "cloud_relay" | "webrtc_turn";
    availability: "available" | "unavailable";
    reason?: string;
}

export interface LiveNetworkCapabilityReport {
    protocolVersion: typeof LIVE_NETWORK_PROTOCOL_VERSION;
    configuredScope: LiveNetworkScope;
    baseUrlOrigin?: string;
    remoteSurfaceEnabled: boolean;
    lanIndependent: boolean;
    nonLoopbackTlsRequired: boolean;
    nonLoopbackDeviceAuthRequired: boolean;
    latencyTelemetry: "viewer_websocket_ping_round_trip";
    inputPolicy: "rtt_adaptive_pointer_coalescing";
    transports: LiveNetworkTransportCapability[];
    privacy: {
        cloudFramePersistence: false;
        cloudOcrPersistence: false;
        telemetryEnabled: false;
        credentialsExposed: false;
        policy: "local_first_no_cloud_storage";
    };
}

export const unavailableLiveNetworkReport: LiveNetworkCapabilityReport = {
    protocolVersion: LIVE_NETWORK_PROTOCOL_VERSION,
    configuredScope: "unavailable",
    remoteSurfaceEnabled: false,
    lanIndependent: true,
    nonLoopbackTlsRequired: true,
    nonLoopbackDeviceAuthRequired: true,
    latencyTelemetry: "viewer_websocket_ping_round_trip",
    inputPolicy: "rtt_adaptive_pointer_coalescing",
    transports: [
        { id: "websocket_binary", availability: "unavailable", reason: "native_runtime_unavailable" },
        { id: "cloud_relay", availability: "unavailable", reason: "relay_provider_not_configured" },
        { id: "webrtc_turn", availability: "unavailable", reason: "nat_traversal_not_configured" },
    ],
    privacy: {
        cloudFramePersistence: false,
        cloudOcrPersistence: false,
        telemetryEnabled: false,
        credentialsExposed: false,
        policy: "local_first_no_cloud_storage",
    },
};

export function parseLiveNetworkCapabilityReport(input: unknown): LiveNetworkCapabilityReport {
    const report = record(input, "live network report");
    exactKeys(report, [
        "protocolVersion", "configuredScope", "remoteSurfaceEnabled", "lanIndependent",
        "nonLoopbackTlsRequired", "nonLoopbackDeviceAuthRequired", "latencyTelemetry",
        "inputPolicy", "transports", "privacy",
    ], ["baseUrlOrigin"]);
    if (report.protocolVersion !== LIVE_NETWORK_PROTOCOL_VERSION) throw new Error("unsupported live network protocol");
    member(report.configuredScope, ["loopback_http", "loopback_https", "private_https", "unavailable"], "scope");
    for (const key of [
        "remoteSurfaceEnabled", "lanIndependent", "nonLoopbackTlsRequired", "nonLoopbackDeviceAuthRequired",
    ]) {
        if (typeof report[key] !== "boolean") throw new Error(`invalid live network ${key}`);
    }
    if (report.lanIndependent !== true || report.nonLoopbackTlsRequired !== true
        || report.nonLoopbackDeviceAuthRequired !== true) {
        throw new Error("live network security posture cannot be weakened");
    }
    if (report.latencyTelemetry !== "viewer_websocket_ping_round_trip"
        || report.inputPolicy !== "rtt_adaptive_pointer_coalescing") {
        throw new Error("invalid live network latency policy");
    }
    validateOrigin(report.baseUrlOrigin, report.configuredScope as LiveNetworkScope);
    if (!Array.isArray(report.transports) || report.transports.length !== 3) {
        throw new Error("invalid live network transport inventory");
    }
    const transports = report.transports.map(parseTransport);
    const ids = new Set(transports.map((transport) => transport.id));
    if (ids.size !== 3 || !["websocket_binary", "cloud_relay", "webrtc_turn"].every((id) => ids.has(id as LiveNetworkTransportCapability["id"]))) {
        throw new Error("incomplete or duplicate live network transport inventory");
    }
    const privacy = parsePrivacy(report.privacy);
    return { ...(report as unknown as LiveNetworkCapabilityReport), transports, privacy };
}

function parseTransport(input: unknown): LiveNetworkTransportCapability {
    const value = record(input, "live network transport");
    exactKeys(value, ["id", "availability"], ["reason"]);
    member(value.id, ["websocket_binary", "cloud_relay", "webrtc_turn"], "transport id");
    member(value.availability, ["available", "unavailable"], "transport availability");
    if (value.reason !== undefined && (typeof value.reason !== "string" || !/^[a-z][a-z0-9_]{0,159}$/.test(value.reason))) {
        throw new Error("invalid live network transport reason");
    }
    if (value.availability === "unavailable" && value.reason === undefined) {
        throw new Error("unavailable live network transport requires a reason");
    }
    if (value.availability === "available" && value.reason !== undefined) {
        throw new Error("available live network transport cannot have a denial reason");
    }
    return value as unknown as LiveNetworkTransportCapability;
}

function parsePrivacy(input: unknown): LiveNetworkCapabilityReport["privacy"] {
    const value = record(input, "live network privacy posture");
    exactKeys(value, [
        "cloudFramePersistence", "cloudOcrPersistence", "telemetryEnabled", "credentialsExposed", "policy",
    ]);
    if (value.cloudFramePersistence !== false || value.cloudOcrPersistence !== false
        || value.telemetryEnabled !== false || value.credentialsExposed !== false
        || value.policy !== "local_first_no_cloud_storage") {
        throw new Error("invalid live network privacy posture");
    }
    return value as unknown as LiveNetworkCapabilityReport["privacy"];
}

function validateOrigin(input: unknown, scope: LiveNetworkScope): void {
    if (scope === "unavailable") {
        if (input !== undefined) throw new Error("unavailable live network scope cannot expose an origin");
        return;
    }
    if (typeof input !== "string" || input.length > 2048) throw new Error("invalid live network origin");
    const url = new URL(input);
    if (url.username || url.password || url.pathname !== "/" || url.search || url.hash) {
        throw new Error("live network endpoint must be an origin without credentials");
    }
    const hostname = url.hostname.toLowerCase().replace(/^\[|\]$/g, "");
    const ipv4 = hostname.split(".").map(Number);
    const loopback = hostname === "localhost" || hostname === "::1"
        || (ipv4.length === 4 && ipv4[0] === 127
            && ipv4.every((part) => Number.isInteger(part) && part >= 0 && part <= 255));
    const matchesScope = scope === "loopback_http"
        ? url.protocol === "http:" && loopback
        : scope === "loopback_https"
            ? url.protocol === "https:" && loopback
            : url.protocol === "https:" && !loopback;
    if (!matchesScope) {
        throw new Error("live network origin does not match its configured scope");
    }
}

function record(input: unknown, name: string): Record<string, unknown> {
    if (typeof input !== "object" || input === null || Array.isArray(input)) throw new Error(`invalid ${name}`);
    return input as Record<string, unknown>;
}

function exactKeys(value: Record<string, unknown>, required: string[], optional: string[] = []): void {
    const allowed = new Set([...required, ...optional]);
    if (required.some((key) => !(key in value)) || Object.keys(value).some((key) => !allowed.has(key))) {
        throw new Error("invalid live network fields");
    }
}

function member(input: unknown, values: readonly string[], name: string): void {
    if (typeof input !== "string" || !values.includes(input)) throw new Error(`invalid live network ${name}`);
}

File: src-tauri/src/native/live_network_capabilities.rs
// Non-secret transport and privacy posture exposed to the live collaboration UI.
const LIVE_NETWORK_PROTOCOL_VERSION: &str = "hook.live.network.v1";

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
enum LiveNetworkScope {
    LoopbackHttp,
    LoopbackHttps,
    PrivateHttps,
    Unavailable,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
enum LiveNetworkAvailability {
    Available,
    Unavailable,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
struct LiveNetworkTransportCapability {
    id: String,
    availability: LiveNetworkAvailability,
    #[serde(skip_serializing_if = "Option::is_none")]
    reason: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
struct LiveNetworkPrivacyPosture {
    cloud_frame_persistence: bool,
    cloud_ocr_persistence: bool,
    telemetry_enabled: bool,
    credentials_exposed: bool,
    policy: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
struct LiveNetworkCapabilityReport {
    protocol_version: String,
    configured_scope: LiveNetworkScope,
    #[serde(skip_serializing_if = "Option::is_none")]
    base_url_origin: Option<String>,
    remote_surface_enabled: bool,
    lan_independent: bool,
    non_loopback_tls_required: bool,
    non_loopback_device_auth_required: bool,
    latency_telemetry: String,
    input_policy: String,
    transports: Vec<LiveNetworkTransportCapability>,
    privacy: LiveNetworkPrivacyPosture,
}

fn unavailable_network_transport(id: &str, reason: &str) -> LiveNetworkTransportCapability {
    LiveNetworkTransportCapability {
        id: id.to_owned(),
        availability: LiveNetworkAvailability::Unavailable,
        reason: Some(reason.to_owned()),
    }
}

fn live_network_capability_report(base_url: Option<&str>) -> LiveNetworkCapabilityReport {
    let (configured_scope, base_url_origin, websocket) = match base_url.and_then(|value| {
        crate::loom_connector::classify_loom_base_url(value)
            .ok()
            .map(|kind| (value, kind))
    }) {
        Some((value, crate::loom_connector::LoomBaseUrlKind::LoopbackHttp)) => (
            LiveNetworkScope::LoopbackHttp,
            Some(value.to_owned()),
            LiveNetworkTransportCapability {
                id: "websocket_binary".to_owned(),
                availability: LiveNetworkAvailability::Available,
                reason: None,
            },
        ),
        Some((value, crate::loom_connector::LoomBaseUrlKind::LoopbackHttps)) => (
            LiveNetworkScope::LoopbackHttps,
            Some(value.to_owned()),
            LiveNetworkTransportCapability {
                id: "websocket_binary".to_owned(),
                availability: LiveNetworkAvailability::Available,
                reason: None,
            },
        ),
        Some((value, crate::loom_connector::LoomBaseUrlKind::RemoteHttps))
            if cfg!(feature = "remote-surface") =>
        {
            (
                LiveNetworkScope::PrivateHttps,
                Some(value.to_owned()),
                LiveNetworkTransportCapability {
                    id: "websocket_binary".to_owned(),
                    availability: LiveNetworkAvailability::Available,
                    reason: None,
                },
            )
        }
        _ => (
            LiveNetworkScope::Unavailable,
            None,
            unavailable_network_transport("websocket_binary", "loom_manifest_unavailable"),
        ),
    };
    LiveNetworkCapabilityReport {
        protocol_version: LIVE_NETWORK_PROTOCOL_VERSION.to_owned(),
        configured_scope,
        base_url_origin,
        remote_surface_enabled: cfg!(feature = "remote-surface"),
        lan_independent: true,
        non_loopback_tls_required: true,
        non_loopback_device_auth_required: true,
        latency_telemetry: "viewer_websocket_ping_round_trip".to_owned(),
        input_policy: "rtt_adaptive_pointer_coalescing".to_owned(),
        transports: vec![
            websocket,
            unavailable_network_transport("cloud_relay", "relay_provider_not_configured"),
            unavailable_network_transport("webrtc_turn", "nat_traversal_not_configured"),
        ],
        privacy: LiveNetworkPrivacyPosture {
            cloud_frame_persistence: false,
            cloud_ocr_persistence: false,
            telemetry_enabled: false,
            credentials_exposed: false,
            policy: "local_first_no_cloud_storage".to_owned(),
        },
    }
}

#[tauri::command]
fn get_live_network_capabilities() -> LiveNetworkCapabilityReport {
    let base_url = crate::loom_connector::read_default_loom_manifest()
        .ok()
        .map(|manifest| manifest.transport.base_url);
    live_network_capability_report(base_url.as_deref())
}

#[cfg(test)]
mod live_network_capability_tests {
    use super::*;

    #[test]
    fn local_and_private_https_scopes_keep_cloud_transports_honest() {
        let local = live_network_capability_report(Some("http://127.0.0.1:8765"));
        assert_eq!(local.configured_scope, LiveNetworkScope::LoopbackHttp);
        assert!(local.lan_independent);
        assert_eq!(
            local.transports[0].availability,
            LiveNetworkAvailability::Available
        );
        assert!(local.transports[1..].iter().all(|transport| {
            transport.availability == LiveNetworkAvailability::Unavailable
                && transport
                    .reason
                    .as_deref()
                    .is_some_and(|reason| !reason.is_empty())
        }));

        let remote = live_network_capability_report(Some("https://loom.example.test"));
        let expected_scope = if cfg!(feature = "remote-surface") {
            LiveNetworkScope::PrivateHttps
        } else {
            LiveNetworkScope::Unavailable
        };
        assert_eq!(remote.configured_scope, expected_scope);
        assert!(remote.non_loopback_tls_required);
        assert!(remote.non_loopback_device_auth_required);
    }

    #[test]
    fn privacy_posture_never_claims_cloud_storage_or_telemetry() {
        let report = live_network_capability_report(None);
        assert_eq!(report.configured_scope, LiveNetworkScope::Unavailable);
        assert!(!report.privacy.cloud_frame_persistence);
        assert!(!report.privacy.cloud_ocr_persistence);
        assert!(!report.privacy.telemetry_enabled);
        assert!(!report.privacy.credentials_exposed);
    }
}
